src/valid/handles.rs

```diff
@@ -0,0 +1,631 @@
+    ///
+    /// # Errors
+    ///
+    /// Errors returned by this method are intentionally sparse, for simplicity of implementation.
+    /// It is expected that only buggy frontends or fuzzers should ever emit IR that fails this
+    /// validation pass.
+    pub(super) fn validate_module_handles(module: &crate::Module) -> Result<(), ValidationError> {
+        let &crate::Module {
+            ref constants,
+            ref entry_points,
+            ref functions,
+            ref global_variables,
+            ref types,
+        } = module;
+
+        // NOTE: Types being first is important. All other forms of validation depend on this.
+        for (this_handle, ty) in types.iter() {
+            let &crate::Type {
+                ref name,
+                ref inner,
+            } = ty;
+
+            let validate_array_size = |size| {
+                match size {
+                    crate::ArraySize::Constant(constant) => {
+                        let &crate::Constant {
+                            name: _,
+                            specialization: _,
+                            ref inner,
+                        } = constants.try_get(constant)?;
+                        if !matches!(inner, &crate::ConstantInner::Scalar { .. }) {
+                            return Err(ValidationError::Type {
+                                handle: this_handle,
+                                name: name.clone().unwrap_or_default(),
+                                source: TypeError::InvalidArraySizeConstant(constant),
+                            });
+                        }
+                    }
+                    crate::ArraySize::Dynamic => (),
+                };
+                Ok(this_handle)
+            };
+
+            match inner {
+                &crate::TypeInner::Scalar { .. }
+                | &crate::TypeInner::Vector { .. }
+                | &crate::TypeInner::Matrix { .. }
+                | &crate::TypeInner::ValuePointer { .. }
+                | &crate::TypeInner::Atomic { .. }
+                | &crate::TypeInner::Image { .. }
+                | &crate::TypeInner::Sampler { .. } => (),
+                &crate::TypeInner::Pointer { base, space: _ } => {
+                    this_handle.check_dep(base)?;
+                }
+                &crate::TypeInner::Array {
+                    base,
+                    size,
+                    stride: _,
+                } => {
+                    this_handle.check_dep(base)?;
+                    validate_array_size(size)?;
+                }
+                &crate::TypeInner::Struct {
+                    ref members,
+                    span: _,
+                } => {
+                    this_handle.check_dep_iter(members.iter().map(|m| m.ty))?;
+                }
+                &crate::TypeInner::BindingArray { base, size } => {
+                    this_handle.check_dep(base)?;
+                    validate_array_size(size)?;
+                }
```

Comment:
Suggestion: combine this `match` arm with with `TypeInner::Array` via an `|` pattern.

Reply:
Changed as suggested with b43468a6.